src/Forms/Validator/index.js

```diff
@@ -0,0 +1,98 @@
+import { Component } from 'react';
+import PropTypes from 'prop-types';
+import { isEmpty, isEqual } from 'lodash/fp';
+import { getErrors } from './service';
+
+const VALIDATION_PROP_TYPES = {
+  required: PropTypes.bool,
+  regex: PropTypes.object,
+  caseInsensitive: PropTypes.bool,
+  allowSpaces: PropTypes.bool,
+  min: PropTypes.number,
+  max: PropTypes.number,
+  values: PropTypes.array
+};
+
+class Validator extends Component {
+  static propTypes = {
+    validations: PropTypes.shape(VALIDATION_PROP_TYPES),
+    fetchRulesFn: PropTypes.func,
+    metaPath: PropTypes.string,
+    value: PropTypes.any,
+    country: PropTypes.string,
+    onValidate: PropTypes.func
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = { meta: {} };
+  }
+
+  componentDidMount() {
+    this.getRules();
+  }
+
+  shouldComponentUpdate(nextProps) {
+    const { value, validations, country } = this.props;
+    const {
+      value: nextValue,
+      validations: nextValidations,
+      country: nextCountry
+    } = nextProps;
+
+    return (
+      nextValue !== value ||
+      nextCountry !== country ||
+      !isEqual(validations, nextValidations)
+    );
+  }
+
+  componentDidUpdate(prevProps) {
+    const { country, validations, value } = this.props;
+    const { country: prevCountry, validations: prevValidations } = prevProps;
+
+    if (country !== prevCountry || !isEqual(validations, prevValidations)) {
+      this.getRules();
+      return;
+    }
+
+    this.evaluateRules(value);
+  }
+
+  evaluateRules(value) {
+    const { meta } = this.state;
+    const { onValidate } = this.props;
+
+    if (isEmpty(meta)) {
+      return;
+    }
+
+    const errors = getErrors(meta, value);
+    onValidate({ meta, errors });
+  }
+
+  getRules() {
+    const { metaPath, validations, value, country, fetchRulesFn } = this.props;
+    if (!metaPath) {
+      this.setAndEvaluate(validations, value);
+      return;
+    }
+
+    // Otherwise, fetch validations from the server
+    fetchRulesFn(metaPath, country).then(meta => {
```

Comment:
TODO: Need to allow this function to be provided -- perhaps the better solution would be to have two different validators, one for standard validations and one for async ones?

Reply:
Cool idea. Having two components splits up the concerns more nicely. Solves the problem of having this weird `meta` property in the form configuration. Right now, merging the response from the endpoint with what we have in `validations` feels really weird.

Not sure, if having two components running validations and updating the form state would cause problems with state or performance, though.